File: 剑指Offer/31.栈的压入弹出序列.cpp
class Solution {
public:
    bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
        if (popped.size() == 0 && pushed.size() == 0) return true;
        if (popped.size() == 0 || pushed.size() == 0) return false;
        
        int index = 0;
        stack<int> s;
        for (int i : pushed) {
            s.push(i);
            while (!s.empty() && s.top() == popped[index]) {
                s.pop();
                index++;
            }
        }
        return index == popped.size();
    }
};

File: 剑指Offer/31.栈的压入弹出序列.py
class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        if len(pushed) == 0 and len(popped) == 0:
            return True
        if len(pushed) == 0 or len(popped) == 0:
            return False

        s = []
        index = 0
        for item in pushed:
            s.append(item)
            while len(s) > 0 and s[-1] == popped[index]:
                s.pop()
                index += 1
        
        return index == len(popped)